source.go

```diff
@@ -288,12 +323,31 @@ func WithCreateDestPath() llb.CopyOption {
 	})
 }
 
-func SourceIsDir(src Source) bool {
+func (s *Source) IsDir(srcState llb.State, sOpt SourceOpts) bool {
+	return SourceIsDir(*s, srcState, sOpt)
+}
+
+func contextSourceIsDir(srcState llb.State, sOpt SourceOpts) bool {
+	srcFS := sOpt.GetFS(srcState)
+	entries, err := srcFS.ReadDir("/")
+	if err != nil {
+		panic(err)
+	}
+	if len(entries) == 1 {
+		return entries[0].IsDir()
+	} else if len(entries) == 0 {
+		panic("shouldn't be the case") // Maybe this could raise a warning instead
```

Comment:
I don't think we should panic here. If someone includes an empty context, it's probably an accident, but we should not assume anything. We could definitely give the user a warning if they include empty sources generally, but I'm not sure this is the place for it (perhaps we could add another piece of validation after we've generated the llb for all sources)

Reply:
I'd say if `len(entries) == 0` then this is a directory.